src/winstanley/wdl.bnf

```diff
@@ -50,13 +50,16 @@ runtime_block ::= RUNTIME map
 parameter_meta_block ::= PARAMETER_META map
 meta_block ::= META map
 
-declaration ::= type_e postfix_quantifier? IDENTIFIER setter?
+declaration ::= type_e IDENTIFIER setter?
 setter ::= EQUAL expression
-postfix_quantifier ::= QMARK | PLUS
 
 map ::= LBRACE kv* RBRACE
 kv ::= IDENTIFIER COLON expression
-type_e ::= TYPE (LSQUARE type_e (COMMA type_e)* RSQUARE)?
+
+type_e ::= (OBJECT_TYPE | PRIMITIVE_TYPE | map_expression | array_expression | pair_expression ) QMARK?
+array_expression ::= ARRAY_TYPE LSQUARE type_e RSQUARE PLUS?
+map_expression ::= MAP_TYPE LSQUARE type_e COMMA type_e RSQUARE
+pair_expression ::= PAIR_TYPE LSQUARE type_e COMMA type_e RSQUARE
```

Comment:
Since the term `expression` is a little bit overloaded, I'd just call these `array_type`, etc.